Check JKS/PKCS12 truststore in Secrets only if issuer provides the CA

vinzent opened the pull request:
### Pull Request Motivation

The current policy check for keystores in Secrets creates a loop because
the truststore.jks or truststore.p12 will never exist when the issuer didn't
provide the CA certificate. This behaviour was introduced by #5597

The JKS and PKCS12 truststores are only added to the Secret
if the CA is provided by the issuer. The CertificateRequest API
reference states:

> The PEM encoded x509 certificate of the signer, also known
> as the CA (Certificate Authority). This is set on a best-effort basis by
> different issuers. If not set, the CA is assumed to be unknown/not available.


This change will only check the PKCS12/JKS truststores if the CA cert from the
issuer exists in the secret.


Fixes #5755

### Kind

/kind bug

### Release Note

```release-note
 Check JKS/PKCS12 truststore in Secrets only if issuer provides the CA
```

vinzent commented:
Getting `invalid memory address or nil pointer dereference`.

vinzent commented:
Seems `input.CurrentRevisionRequest` and `input.NewRevisionRequest` is always nil/null. So can't check this to verify if there is a CA or not.

speer commented:
> Seems `input.CurrentRevisionRequest` and `input.NewRevisionRequest` is always nil/null. So can't check this to verify if there is a CA or not.

Would it make sense then to remove the validation for the truststore completely, i.e., removing the second part from the condition?

if len(input.Secret.Data[cmapi.PKCS12SecretKey]) == 0 ~~|| len(input.Secret.Data[cmapi.PKCS12TruststoreKey]) == 0~~ {

vinzent commented:
FYI @sathyanarays

irbekrm reviewed (changes requested):
Thanks for the PR @vinzent !

Generally looks good to me, but I think we should run the check if  the CA cert exists

vinzent commented on internal/controller/certificates/policies/checks.go:
```diff
@@ -112,10 +112,11 @@ func SecretKeystoreFormatMatchesSpec(input Input) (string, string, bool) {
+			if len(input.Secret.Data[cmapi.JKSSecretKey]) == 0 {
+				return SecretMismatch, "JKS Keystore key does not contain data", true
 			}
+			// Can't check the Secrets truststore.jks (cmapi.JKSTruststoreKey) as
```
at the check, we only have the cmapi.Certificate data and the corev1.Secret. So IMHO I can only check like: 

```go
if input.Secret.Data[cmapi.PKCS12TruststoreKey] != nil &&
    len(input.Secret.Data[cmapi.PKCS12TruststoreKey]) == 0 {
    return SecretMismatch, "PKCS12 Truststore key does not contain data", true
}
```

This will only check if when it already exists, then it should not be empty. IMHO this has not much value? But I still don't know if the CA needs to exist or not.  There is no check like `input.Certificate.Spec.Keystores.JKS.Create` to know whether the CA needs to exist or not.

irbekrm replied:
> This will only check if when it already exists, then it should not be empty. IMHO this has not much value

I think it has the same value as all the other checks that ensure that if, for example, the secret was manually modified in the cluster, those changes will be overriden.

irbekrm replied:
> the corev1.Secret

The Secret has CA if the issuer returns a CA cert and does not have it if it does not, so you should be able to rely on that I think

irbekrm replied:
I think this is what we should do

```
func SecretKeystoreFormatMatchesSpec(input Input) (string, string, bool) {
...
_, issuerProvidesCA := input.Secret.Data[cmmeta.TLSCAKey] // check that this issuer provides CA
...
	if input.Certificate.Spec.Keystores.JKS != nil {
		if input.Certificate.Spec.Keystores.JKS.Create {
			if len(input.Secret.Data[cmapi.JKSSecretKey]) == 0 ||
				(len(input.Secret.Data[cmapi.JKSTruststoreKey]) == 0 && issuerProvidesCA) {
				return SecretMismatch, "JKS Keystore keys does not contain data", true
			}
...
```

vinzent replied:
I think I got it know.  :-) Thanks. Will update the code.

wallrj reviewed (commented):
Sorry to weigh in on this. But would it be a valid alternative approach to *only* check for the **existence** of the `.p12` files.
And unconditionally create the `truststore.p12` file (maybe empty) regardless of whether there is a corresponding `ca.crt` file.
Then the checks would be simpler.

Downside is that it might not be backwards compatible, because suddenly users would see empty truststore.p12 files in their mounted Secrets, where previously they only attempted to load them if present.

Please ignore this suggestion if your prefer the  approach that has already been discussed.

wallrj commented:
/ok-to-test

wallrj commented on pkg/apis/certmanager/v1/types_certificate.go:
Need regenerate the CRD files by runing `make update-crds` whenever the API types change.
See 
 * https://prow.build-infra.jetstack.net/view/gs/jetstack-logs/pr-logs/pull/cert-manager_cert-manager/5972/pull-cert-manager-master-make-test/1651590104839360512

irbekrm replied:
This is also not ready yet https://github.com/cert-manager/cert-manager/pull/5972#discussion_r1178059723

irbekrm commented:
> Then the checks would be simpler.
Downside is that it might not be backwards compatible, because suddenly users would see empty truststore.p12 files in their mounted Secrets, where previously they only attempted to load them if present.

I don't think the check is that complex and I think that the one extra comparison is preferable to adding an empty truststore.p12- I wouldn't care that the users _see_ it, but perhaps have something that reads it.

wallrj commented:
> Downside is that it might not be backwards compatible, because suddenly users would see empty truststore.p12 files in their mounted Secrets, where previously they only attempted to load them if present.

I think I've talked myself out of that idea. Ignore it.

vinzent commented:
/retest

vinzent commented:
@irbekrm i've implemented your suggestion, rebased to latest master and ran the `make update-crds` @wallrj suggested. I have not yet verified that it actually works in my cluster. :)

wallrj commented on internal/controller/certificates/policies/checks.go:
Another drive by comment.
The original PR added tests which demonstrated the Secret being updated. Can those original tests be adjusted to demonstrate the new correct behaviour. 
 * https://github.com/cert-manager/cert-manager/pull/5597/files#diff-45cc1a739562a5093d3599b72afae24bc276569c97d02399d4f52e9db40b4975

And in hindsight I think those tests were too high level. 
It would be sufficient to add some unit-tests called `TestSecretKeystoreFormatMatchesSpec`,
in: 
 * https://github.com/cert-manager/cert-manager/blob/master/internal/controller/certificates/policies/checks_test.go

wallrj reviewed (changes requested):
@vinzent Do you have time to add some tests for this change?

irbekrm reviewed (approved):
Thanks for working on this and fixing the bug @vinzent 

I've tested this with a cert with jks keystores defined for both a selfsigned issuer that provides a CA and with LetsEncrypt staging issuer that does not - in the first case the `truststore.jks` got populated as expected, in the second case it did not and cert-manager was also not throwing any errors, so this seems fixed.

Adding a hold in case you have time to add tests that Richard mentions, but also aware that we cannot expect a bugfix to improve tests from a different PR. I think this should go into v1.12 in any case

/hold
/lgtm

irbekrm commented:
/milestone 1.12

vinzent commented:
@irbekrm I built the image and verified it on my cluster. Tested with pkcs12. Doesn't loop anymore on the "missing" truststore.p12.

irbekrm commented:
I will create a separate issue for tests

/hold cancel
